atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1404 @@
+    ///         node_small_id: 1,
+    ///     };
+    ///     let keystore = Arc::new(FileBasedKeystore::new(&std::path::PathBuf::from("keys"))?);
+    ///     
+    ///     AtomaP2pNode::start(config, keystore, state_manager_sender, false)
+    /// }
+    /// ```
+    ///
+    /// # Network Architecture
+    ///
+    /// The node uses a layered networking approach:
+    /// 1. Transport Layer: TCP/QUIC with noise encryption
+    /// 2. Protocol Layer:
+    ///    - Gossipsub for message broadcasting
+    /// 3. Application Layer: Custom message handling for node addresses and requests
+    ///
+    /// # Security Considerations
+    ///
+    /// - All network connections are encrypted using the noise protocol
+    /// - Messages are signed using the node's private key
+    /// - Peer connections are authenticated
+    /// - The node validates all incoming messages
+    #[instrument(level = "info", skip_all)]
+    pub fn start(
+        config: AtomaP2pNodeConfig,
+        keystore: Arc<FileBasedKeystore>,
+        state_manager_sender: Sender<StateManagerEvent>,
+        is_client: bool,
+    ) -> Result<Self, AtomaP2pNodeError> {
+        let mut swarm = SwarmBuilder::with_new_identity()
+            .with_tokio()
+            .with_quic()
+            .with_behaviour(|key| {
+                // To content-address message, we can take the hash of message and use it as an ID.
+                let message_id_fn = |message: &gossipsub::Message| {
+                    let mut s = DefaultHasher::new();
+                    message.data.hash(&mut s);
+                    gossipsub::MessageId::from(s.finish().to_string())
+                };
+
+                let gossipsub_config = gossipsub::ConfigBuilder::default()
+                    .heartbeat_interval(config.heartbeat_interval)
+                    .validate_messages()
+                    .validation_mode(gossipsub::ValidationMode::Strict)
+                    .message_id_fn(message_id_fn)
+                    .build()
+                    .map_err(|e| {
+                        error!(
+                            target = "atoma-p2p",
+                            event = "build_gossipsub_config",
+                            error = %e,
+                            "Failed to build gossipsub config"
+                        );
+                        AtomaP2pNodeError::GossipsubConfigError(e)
+                    })?;
+
+                let gossipsub = gossipsub::Behaviour::new(
+                    gossipsub::MessageAuthenticity::Signed(key.clone()),
+                    gossipsub_config,
+                )?;
+
+                let store = kad::store::MemoryStore::new(key.public().to_peer_id());
+                let kademlia = kad::Behaviour::new(key.public().to_peer_id(), store);
+
+                let mdns = mdns::tokio::Behaviour::new(
+                    mdns::Config::default(),
+                    key.public().to_peer_id(),
+                )?;
+
```

Comment:
We should also include the [identify behaviour](https://github.com/libp2p/rust-libp2p/blob/56691dbc33de82e221da14ccea5edc04074e94c4/examples/identify/README.md#L4) as a part of the config